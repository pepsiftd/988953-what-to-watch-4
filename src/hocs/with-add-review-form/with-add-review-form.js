import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {ReviewModel} from '@/models/review-model';

const InputName = {
  'rating': `rating`,
  'review-text': `comment`,
};

const COMMENT_MIN_LENGTH = 50;
const COMMENT_MAX_LENGTH = 400;

const Error = {
  NO_RATING: `Please rate the movie by selecting stars amount`,
  COMMENT_OUT_OF_RANGE: `Your review should not be shorter than ${COMMENT_MIN_LENGTH} or longer than ${COMMENT_MAX_LENGTH} symbols`
};

const withAddReviewForm = (Component) => {
  class WithAddReviewForm extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        rating: null,
        comment: ``,
        isFormDisabled: false,
        isSubmitButtonDisabled: false,
        errors: [],
      };

      this.onSubmit = this.onSubmit.bind(this);
      this.onInput = this.onInput.bind(this);
    }

    _setError() {
      const errors = [];
      if (this.state.rating === null) {
        errors.push(Error.NO_RATING);
      }
      if (this.state.comment.length < 50 || this.state.comment.length > 400) {
        errors.push(Error.COMMENT_OUT_OF_RANGE);
      }
      return this.setState({
        errors,
      });
    }

    _validateForm() {
      const {rating, comment} = this.state;

      return this.setState({
        isSubmitButtonDisabled: !rating || comment.length < COMMENT_MIN_LENGTH || comment.length > COMMENT_MAX_LENGTH
      });
    }

    onInput(evt) {
      const {name, value} = evt.target;

      this.setState({
        [InputName[name]]: value
      }, () => {
        if (this.state.isSubmitButtonDisabled) {
          this._validateForm();
          this._setError();
        }
      });
    }

    onSubmit() {
      Promise.all([this._validateForm(), this._setError()])
        .then(() => {
          const id = this.props.movie.id;
          const review = ReviewModel.toRAW(this.state.rating, this.state.comment);

          if (this.state.errors.length === 0) {
            this.props.onSendReview(id, review);
          }
        });
    }

    render() {
      const {isFormDisabled, isSubmitButtonDisabled, errors} = this.state;
      return (
        <Component
          {...this.props}
          isFormDisabled={isFormDisabled}
          isSubmitButtonDisabled={isSubmitButtonDisabled}
          onSubmit={this.onSubmit}
          onInput={this.onInput}
          errors={errors}
        />
      );
    }
  }

  WithAddReviewForm.propTypes = {
    movie: PropTypes.shape({
      id: PropTypes.number.isRequired,
    }).isRequired,
    onSendReview: PropTypes.func.isRequired,
  };

  return WithAddReviewForm;
};

export {withAddReviewForm};
