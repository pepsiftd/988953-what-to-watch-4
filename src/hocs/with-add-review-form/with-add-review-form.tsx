import React, {PureComponent} from 'react';
import {Subtract} from 'utility-types';

import {ReviewModel} from '@/models/review-model';
import {COMMENT_MIN_LENGTH, COMMENT_MAX_LENGTH} from '@/const';

const InputName = {
  'rating': `rating`,
  'review-text': `comment`,
};

const Error = {
  NO_RATING: `Please rate the movie by selecting stars amount`,
  COMMENT_OUT_OF_RANGE: `Your review should not be shorter than ${COMMENT_MIN_LENGTH} or longer than ${COMMENT_MAX_LENGTH} symbols`
};

interface State {
  rating: number;
  comment: string;
  isFormDisabled: boolean;
  isSubmitButtonDisabled: boolean;
  errors: string[];
}

interface InjectingProps {
  isFormDisabled: boolean;
  isSubmitButtonDisabled: boolean;
  onSubmit: () => void;
  onInput: (evt: React.FormEvent) => void;
  errors: string[];
}

const withAddReviewForm: (Component: React.ComponentClass | React.FunctionComponent) => React.ComponentClass = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithAddReviewForm extends PureComponent<T, State> {
    constructor(props) {
      super(props);

      this.state = {
        rating: null,
        comment: ``,
        isFormDisabled: false,
        isSubmitButtonDisabled: true,
        errors: [],
      };

      this.onSubmit = this.onSubmit.bind(this);
      this.onInput = this.onInput.bind(this);
    }

    _setError(errorText = null) {
      const errors = [];
      if (this.state.rating === null) {
        errors.push(Error.NO_RATING);
      }
      if (this.state.comment.length < COMMENT_MIN_LENGTH || this.state.comment.length > COMMENT_MAX_LENGTH) {
        errors.push(Error.COMMENT_OUT_OF_RANGE);
      }
      if (errorText) {
        errors.push(errorText);
      }
      return this.setState({
        errors,
      });
    }

    _validateForm() {
      const {rating, comment} = this.state;

      return this.setState({
        isSubmitButtonDisabled: !rating || comment.length < COMMENT_MIN_LENGTH || comment.length > COMMENT_MAX_LENGTH
      });
    }

    onInput(evt) {
      const {name, value} = evt.target;

      this.setState({
        [InputName[name]]: value
      }, () => {
        this._validateForm();
        this._setError();
      });
    }

    onSubmit() {
      Promise.all([this._validateForm(), this._setError()])
        .then(() => {
          const id = this.props.movie.id;
          const review = ReviewModel.toRAW(this.state.rating, this.state.comment);
          const handleSendReviewError = (err) => {
            this._setError(err.message);
            this.setState({
              isFormDisabled: false,
            });
          };

          const handleSendReviewSuccess = () => {
            this.setState({
              isFormDisabled: false,
            });
          };

          if (this.state.errors.length === 0) {
            this.setState({
              isFormDisabled: true,
            });
            this.props.onSendReview(id, review, handleSendReviewSuccess, handleSendReviewError);
          }
        });
    }

    render() {
      const {isFormDisabled, isSubmitButtonDisabled, errors} = this.state;
      return (
        <Component
          {...this.props}
          isFormDisabled={isFormDisabled}
          isSubmitButtonDisabled={isSubmitButtonDisabled}
          onSubmit={this.onSubmit}
          onInput={this.onInput}
          errors={errors}
        />
      );
    }
  }

  return WithAddReviewForm;
};

export {withAddReviewForm};
