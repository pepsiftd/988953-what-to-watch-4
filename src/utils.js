const getUniqueItems = (list) => {
  return list.filter((it, i, array) => {
    return array.indexOf(it) === i;
  });
};

const extend = (a, b) => {
  return Object.assign({}, a, b);
};

const getMoviesByGenre = (movies, genre) => {
  return genre === `All genres` ? movies : movies.filter((movie) => movie.genre === genre);
};

const getGenresFromMovies = (movies) => {
  return getUniqueItems(movies.map((movie) => movie.genre));
};

export {extend, getMoviesByGenre, getUniqueItems, getGenresFromMovies};
