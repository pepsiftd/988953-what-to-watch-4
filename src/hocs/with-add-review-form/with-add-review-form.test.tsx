import React from 'react';
import renderer from 'react-test-renderer';
import PropTypes from 'prop-types';
import {noop} from '@/utils';

import {withAddReviewForm} from './with-add-review-form';

const MockComponent = ({isFormDisabled, isSubmitButtonDisabled, onSubmit, onInput, errors}) => {
  return (
    <form
      onSubmit={onSubmit}
      disabled={isFormDisabled}
    >
      {errors}
      <input type="text" onInput={onInput} />
      <input type="submit" disabled={isSubmitButtonDisabled} />
    </form>
  );
};

MockComponent.propTypes = {
  errors: PropTypes.arrayOf(PropTypes.string).isRequired,
  onInput: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  isSubmitButtonDisabled: PropTypes.bool.isRequired,
  isFormDisabled: PropTypes.bool.isRequired,
};


it(`withAddReviewForm renders correctly`, () => {
  const MockComponentWrapped = withAddReviewForm(MockComponent);

  const tree = renderer.create(
      <MockComponentWrapped
        movie={{
          id: 5
        }}
        onSendReview={noop}
      />
  );

  expect(tree).toMatchSnapshot();
});
