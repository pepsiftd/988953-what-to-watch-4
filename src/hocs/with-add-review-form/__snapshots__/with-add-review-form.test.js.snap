// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withAddReviewForm renders correctly 1`] = `
<form
  disabled={false}
  onSubmit={[Function]}
>
  <input
    onInput={[Function]}
    type="text"
  />
  <input
    disabled={true}
    type="submit"
  />
</form>
`;
